In [19]:
from koapy import KiwoomOpenApiPlusEntrypoint
import pandas as pd
from tqdm import tqdm

In [2]:
entrypoint = KiwoomOpenApiPlusEntrypoint()

2023-04-25 19:24:58,177 [DEBUG] Testing if client is ready... - KiwoomOpenApiPlusEntrypoint.py:45
2023-04-25 19:24:58,191 [DEBUG] Client is ready, using existing server - KiwoomOpenApiPlusEntrypoint.py:55


In [3]:
entrypoint.EnsureConnected()
entrypoint.IsConnected()

True

In [7]:
from koapy import KiwoomOpenApiPlusTrInfo

In [8]:
opt10059_info = KiwoomOpenApiPlusTrInfo.get_trinfo_by_code("opt10059")

In [9]:
opt10059_info

KiwoomOpenApiPlusTrInfo('opt10059', 'STOCKG', '', '1', '2330', '종목별투자자기관별요청', [Field('일자', 0, 20, 9004), Field('종목코드', 20, 20, 9001), Field('금액수량구분', 40, 20, 9036), Field('매매구분', 60, 20, 9032), Field('단위구분', 80, 20, 9207)], '', OutputsParams(None, None, None), [], '종목별투자자기관별', OutputsParams('100', '4', '*'), [Field('일자', 0, 20, 22), Field('현재가', 20, 20, 10), Field('대비기호', 40, 20, 25), Field('전일대비', 60, 20, 11), Field('등락율', 80, 20, 12), Field('누적거래량', 100, 20, 13), Field('누적거래대금', 100, 20, 14), Field('개인투자자', 120, 20, 1859), Field('외국인투자자', 140, 20, 1965), Field('기관계', 160, 20, 1977), Field('금융투자', 180, 20, 1805), Field('보험', 200, 20, 1811), Field('투신', 220, 20, 1817), Field('기타금융', 240, 20, 1835), Field('은행', 260, 20, 1829), Field('연기금등', 280, 20, 1841), Field('사모펀드', 300, 20, 1823), Field('국가', 320, 20, 1847), Field('기타법인', 340, 20, 1853), Field('내외국인', 360, 20, 1971)])

In [10]:
opt10059_info.inputs

[Field('일자', 0, 20, 9004),
 Field('종목코드', 20, 20, 9001),
 Field('금액수량구분', 40, 20, 9036),
 Field('매매구분', 60, 20, 9032),
 Field('단위구분', 80, 20, 9207)]

In [11]:
opt10059_info.single_outputs

[]

In [12]:
opt10059_info.multi_outputs

[Field('일자', 0, 20, 22),
 Field('현재가', 20, 20, 10),
 Field('대비기호', 40, 20, 25),
 Field('전일대비', 60, 20, 11),
 Field('등락율', 80, 20, 12),
 Field('누적거래량', 100, 20, 13),
 Field('누적거래대금', 100, 20, 14),
 Field('개인투자자', 120, 20, 1859),
 Field('외국인투자자', 140, 20, 1965),
 Field('기관계', 160, 20, 1977),
 Field('금융투자', 180, 20, 1805),
 Field('보험', 200, 20, 1811),
 Field('투신', 220, 20, 1817),
 Field('기타금융', 240, 20, 1835),
 Field('은행', 260, 20, 1829),
 Field('연기금등', 280, 20, 1841),
 Field('사모펀드', 300, 20, 1823),
 Field('국가', 320, 20, 1847),
 Field('기타법인', 340, 20, 1853),
 Field('내외국인', 360, 20, 1971)]

In [13]:
import datetime

date = datetime.datetime.now().strftime("%Y%m%d")
date

'20230425'

In [14]:
request_name = "종목별투자자"
tr_code = "opt10059"
screen_no = "0001"
inputs = {
    "일자" : "date",
    "종목코드" : "005930",
    "금액수량구분" : "2",
    "매매구분" : "0",
    "단위구분" : "1000"
}

In [15]:
data_list = []

In [20]:
# 이 코드를 해석해야 내가 원하는 수 만큼 데이터를 받을 수 있겠다
for event in tqdm(entrypoint.TransactionCall(request_name, tr_code, screen_no, inputs)):
    columns = event.multi_data.names
    records = [values.values for values in event.multi_data.values]
    data = pd.DataFrame.from_records(records, columns=columns)
    data_list.append(data)

90it [05:19,  3.55s/it]


In [21]:
pd_data = pd.concat(data_list, axis=0).reset_index(drop=True)
pd_data

일자     현재가 대비기호   전일대비    등락율     누적거래량   누적거래대금  개인투자자 외국인투자자  \
0     20230425  -63600    5  -1600  -2.45  16193271  1039508   1262   -714   
1     20230424  -65200    5   -500  -0.76  12986581   846721  -1631   1358   
2     20230421  +65700    2   +400  +0.61  10538622   691721  -2004   2685   
3     20230420  -65300    5   -200  -0.31   9501169   618181   -829   1258   
4     20230419  -65500    5   -100  -0.15  10255985   672185   -971   1056   
...        ...     ...  ...    ...    ...       ...      ...    ...    ...   
9100  19850109    8010    0    290  +3.62      4970        1      0      0   
9101  19850108    8300    0    110  +1.33     12930        4      0      0   
9102  19850107    8410    0     20  +0.24     11810        3      0      0   
9103  19850105    8390    0     60  +0.72      1660        0      0      0   
9104  19850104    8450    0      0  +0.00      1710        0      0      0   

       기관계  금융투자   보험    투신 기타금융  은행  연기금등  사모펀드 국가 기타법인 내외국인  
0     -543   510   24  -394  -13  21  -311  -380  0  -10    4  
1      341  1152  -18  -392   -9   3  -397     2  0  -62   -7  
2     -670   233   -2  -422   25  -3  -272  -230  0   -3   -8  
3     -404    85    6  -165   -2   3  -494   165  0  -24   -2  
4      -69    91  -46   -60   -4   4  -146    92  0  -11   -5  
...    ...   ...  ...   ...  ...  ..   ...   ... ..  ...  ...  
9100     0     0    0     0    0   0     0     0  0    0    0  
9101     0     0    0     0    0   0     0     0  0    0    0  
9102     0     0    0     0    0   0     0     0  0    0    0  
9103     0     0    0     0    0   0     0     0  0    0    0  
9104     0     0    0     0    0   0     0     0  0    0    0  

[9105 rows x 20 columns]

기록 : 
9105 행을 다 받을 필요는 없다. 나에게는 7년만 있으면 된다 (1680 행)
앞으로 해야 할 일 : 기초 엑셀파일을 종목별로 한 디렉토리 안에 모두 만들어 놓기, 선택한 종목만 다운받기와 모든 종목 다운받기 기능 만들기, 다운 받았으면 디렉토리 안의 엑셀파일의 4페이지에 양식에 맞게 파싱하기, 0797(opt10051) 도 똑같이 만들기